### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by Zophed**
Still need to implement some unit tests, find issues in code, everything has been implemented however.

### CODE REVIEW

**Comment by alexvthornton** [comment]
•	There is an existing BaseLogger class. It needs an auto property to hold the class name. This property should be set in the LogFactory using an object initializer. ❌
•	Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
o	The current date/time ✔
o	The name of the class that created the logger ❌
o	The log level ✔
o	The message ✔
o	The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
•	The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLoggermethod. ❌
•	If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ❌
•	Inside of BaseLoggerMixins implement extension methods on BaseLogger for
o	Error, ✔
o	Warning, ❌ 
o	Information, and ❌
o	Debug. ❌ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
•	Use the nameof() operator when identifying the class name to the logger ❌
•	Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
•	Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
•	Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
•	Set LangVersion to 9.0 ✔
•	Turn on Nullability (Nullable) ✔
•	All of the above should be unit tested.
Extra Credit
•	Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses System.Console or System.Diagnostics.Trace ❌

**Comment by alexvthornton** [comment]
- ClassName is not set with an object initializer. 	
- Hard coded the name of the class that created the logger
- FilePath in LogFactory is not a private member
- Doesn’t return null if configureFileLogger hasn’t been called yet
- In extention methods Warning, Information, and Debug in BaseLoggerMixins all have log levels of Error
- Didn’t use nameOf() to Identify the class to the logger

File: Logger/Logger.csproj [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
```
Comment (mmwoodfo) [inline Logger/Logger.csproj:4]: Looks good. 👍🏻

File: Logger/LogFactory.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using System.IO;
+
+namespace Logger
+{
+    public class LogFactory
+    {
+        private string? FilePath { get; set; }
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:8]:
```suggestion
        private string? _FilePath;
```

File: Logger/LogFactory.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+        public void ConfigureFileLogger(string filePath)
+        {
+            this.FilePath = filePath;
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:12]:
```suggestion
            _FilePath = filePath;
```

File: Logger/LogFactory.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            if (FilePath == null)
+            {
+                return null;
+            }
+            else if (!File.Exists(FilePath))
+            {
+                return null;
+            }
+
+            FileLogger logger = new FileLogger(FilePath) { ClassName = className };
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:30]:
```suggestion
            return new FileLogger(FilePath) { ClassName = className };
```
Just a suggestion. Good job here. :)

File: Logger/LogFactory.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            /*
+             * Cancels the operation if the filePath value was not assignmed by
+             * the ConfigureFileLogger method
+            */
+            if (FilePath == null)
+            {
+                return null;
+            }
+            else if (!File.Exists(FilePath))
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:25]:
```suggestion

```
You can use a StreamWriter to create the file if it does not exist inside the `FileLogger`.

File: Logger/FileLogger.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+        public FileLogger(string filePath)
+        {
+            FilePath = filePath;
+        }
+
+        public override void Log(LogLevel logLevel, string message)
+        {
+            TextWriter appendMessage;
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:24]:
```C#
using (StreamWriter streamWriter = new StreamWriter(_filePath, true))
{
    streamWriter.WriteLine($"{DateTime.Now} {ClassName} {logLevel}: {message}");
}
```
A safer alternative for writing text to a file.

File: Logger/FileLogger.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            if (FilePath != null && File.Exists(FilePath))
+            {
+                appendMessage = new StreamWriter(FilePath);
+                string dateTime = DateTime.Now.ToString("yyyy-MM-dd_HH:mm");
+                appendMessage.WriteLine("Date/time: " + dateTime);
+                appendMessage.WriteLine(base.ClassName);
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:30]: This should all be one line.

File: Logger/FileLogger.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.IO;
+
+namespace Logger
+{
+    public class FileLogger : BaseLogger
+    {
+        private string? filePath;
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:10]:
```suggestion
        private string _FilePath;
```
If you set this in the constructor & check for null there you don't have to mark it as nullable.

File: Logger/FileLogger.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        public string FilePath
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:11]:
```suggestion

```
This is redundant - you already have this defined on line 10.

File: Logger/FileLogger.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        {
+            get => filePath!;
+            set => filePath = value ?? throw new ArgumentNullException(null);
+        }
+
+        public FileLogger(string filePath)
+        {
+            FilePath = filePath;
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:19]:
```suggestion
            _FilePath = filePath ?? throw new ArgumentNullException(nameof(filePath));
```

File: Logger/BaseLoggerMixins.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+{
+    public static class BaseLoggerMixins
+    {
+        public static void Error(this BaseLogger? baseLogger, string message, params object[] arr)
+        {
+            
+             if (baseLogger == null)
+                 throw new ArgumentNullException(nameof(baseLogger));
+
+             baseLogger.Log(LogLevel.Error, string.Format(message, arr[0]));
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:13]:
```suggestion
             baseLogger.Log(LogLevel.Error, string.Format(message, arr));
```

File: Logger/BaseLoggerMixins.cs [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+             
+        }
+
+        public static void Warning(this BaseLogger? baseLogger, string message, params object[] arr)
+        {
+            if (baseLogger == null)
+                throw new ArgumentNullException(nameof(baseLogger));
+
+            baseLogger.Log(LogLevel.Warning, string.Format(message, arr[0]));
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:22]:
```suggestion
            baseLogger.Log(LogLevel.Warning, string.Format(message, arr));
```

File: Logger/BaseLoggerMixins.cs [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        public static void Information(this BaseLogger? baseLogger, string message, params object[] arr)
+        {
+            if (baseLogger == null)
+                throw new ArgumentNullException(nameof(baseLogger));
+
+            baseLogger.Log(LogLevel.Information, string.Format(message, arr[0]));
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:30]:
```suggestion
            baseLogger.Log(LogLevel.Information, string.Format(message, arr));
```

File: Logger/BaseLoggerMixins.cs [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        public static void Debug(this BaseLogger? baseLogger, string message, params object[] arr)
+        {
+
+            if (baseLogger == null)
+                throw new ArgumentNullException(nameof(baseLogger));
+
+            baseLogger.Log(LogLevel.Debug, string.Format(message, arr[0]));
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:39]:
```suggestion
            baseLogger.Log(LogLevel.Debug, string.Format(message, arr));
```

File: Logger.Tests/LogFactoryTests.cs [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+{
+    [TestClass]
+    public class LogFactoryTests
+    {
+        [TestMethod]
+        public void ConfigureFileLogger_FileLoggerNotInstantiated_ReturnsNull()
+        {
+            //Arrange
+            LogFactory logFactory = new LogFactory();
+            BaseLogger? logger;
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:15]:
```suggestion

```
This can just be on line 18.

File: Logger.Tests/LogFactoryTests.cs [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+            //Act
+            logger = logFactory.CreateLogger("");
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:18]:
```suggestion
            BaseLogger? logger = logFactory.CreateLogger(nameof(LogFactoryTests));
```

Be sure to use `nameof` when passing the class name to your logger.

File: Logger.Tests/LogFactoryTests.cs [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+            //Assert
+            Assert.IsNull(logger);
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:21]: 👍🏻

File: Logger.Tests/LogFactoryTests.cs [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        [TestMethod]
+        public void ConfigureFileLogger_PassInBadFilePath_ReturnsNull()
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:25]: Where are you passing in the bad file path?

File: Logger.Tests/LogFactoryTests.cs [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            //Arrange
+            LogFactory logFactory = new LogFactory();
+            BaseLogger? logger;
+
+            //Act
+            logger = logFactory.CreateLogger("BadFile");
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:32]:
```suggestion
            logger = logFactory.CreateLogger(nameof(LogFactoryTests));
```

File: Logger.Tests/LogFactoryTests.cs [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            //Arrange
+            LogFactory? logger = new LogFactory();
+
+
+            //Act
+            logger.CreateLogger(nameof(FileLogger));
+
+            //Assert
+            Assert.AreEqual("FileLogger", nameof(FileLogger));
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:49]: This is not testing your `FileLogger` class, is this intentional?

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+        public void Warning_WithData_LogsMessage()
+        {
+            //Arrange
+            var logger = new TestLogger();
+
+            //Act
+            logger.Warning("Warning Message {0}", 31);
+
+            //Assert
+            Assert.AreEqual("Warning Message 31", logger.LoggedMessages[0].Message);
```
Comment (mmwoodfo) [inline Logger.Tests/BaseLoggerMixinsTests.cs:79]: You are missing some asserts here. Same goes for the tests below.
```suggestion
            Assert.AreEqual("Warning Message 31", logger.LoggedMessages[0].Message);
            Assert.AreEqual(1, logger.LoggedMessages.Count);
            Assert.AreEqual(LogLevel.Warning, logger.LoggedMessages[0].LogLevel);
```

File: Logger.Tests/FileLoggerTests.cs [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+
+            //Assert
+
+        }
+
+        [TestMethod]
+        public void FileLogger_GoodFilePath_LogMatchesInput()
+        {
+            string? filePath;
+            if (File.Exists("testFile.txt"))
```
Comment (mmwoodfo) [inline Logger.Tests/FileLoggerTests.cs:60]: You should create & clean up a file programatically if you intend to use it in a unit test - this test will not pass if run on a computer that does not have the testFile.txt created.

**Review by mmwoodfo — COMMENTED** [review]
##  (Zophed)

## main

| Build/Compile (Build succeeded: 0 Error(s), 0 Warning(s))| **<10>** |
| --------------------- | ------ |
| Assignment compiles even though major pieces might be missing, or there are significant problems: **✔** |  **5** |


| Assignment  | **<10 (missing lots) - 20 (minor)>** |
| --------------------- | ------ |
| <Project Assignment Goes Here> | ❌✔ |


| Project Analysis |   |
| -----------------| - |
| Disabled Warnings **** | **✔**
| Nullable Enabled: | **❌** |
| Appropriately handle all warnings | **✔** |
| CodeReview Performed |  **✔**  |
| Add the code analysis packages and appropriately handle all warnings* |  **❌** |
*<sub><sup>`IntelliTectAnalyzer.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.CodeQuality.Analyzers.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.NetFramework.Analyzers.dll`. (It is okay to appropriately disable rules such as  localization/culture rules.)</sub></sup>


| ProjectName | Nullable | Solutions | Analyzers | Unexpected Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| Logger | Enable | Logger.sln |  |  | 4 |
| Logger.Tests | Enable | Logger.sln |  |  | 4 |
| LoginStuff | Disable |  |  |  | 4 |
| LoginStuff.Tests | Disable |  |  |  | 4 |


| Testing | **<0 - 20>** |
| --------------------- | ------ |
| Test Run FAILED: Total tests=16, Passed=15, Failed=1, Skipped=0 | **70%** |
| Unit test project fails to compile, most test fails, or contains significant problems. | **<10>** |
| Most assigned work is tested or contain some errors. | **<5>** |
| All assigned work is tested | **<5>** |


| Extra Credit | **<0 - 10>** |
| --------------------- | ------ |
| <Extra Credit Goes Here> | ❌✔ |

**Pull request closed without merging**